### PR TITLE
Add support for deep generics resolution

### PR DESCRIPTION
**Pull request by yrodiere**
For instance, given the following model:

```
class A<T extends C> {
  GenericType<T> propertyOfA;
}
class B extends A<D> {
}
class C {
}
class D extends C {
}
class GenericType<T> {
  T propertyOfGenericType;
}
```

We will now model property `B.propertyOfA` as having type `GenericType<D>`
instead of `GenericType<T extends C>` as we used to.

This will also be true for more deeply nested references to a type
variable, for instance the type of property
`B.propertyOfA.propertyOfGenericType` will correctly be inferred as `D`.

### CODE REVIEW

File: mapper-pojo/src/main/java/org/hibernate/search/v6poc/entity/pojo/util/impl/GenericTypeContext.java [hunk 1]
```diff
@@ -23,59 +23,139 @@
...
+ * class B&lt;T extends C&gt; {
+ *   T propertyOfB;
+ * }
+ * class C {
+ * }
+ * class D extends C {
+ * }
+ * </code></pre>
+ * ... creating a context for the type of property {@code propertyOfB}, within the "declaring" context of type {@code A},
+ * will result in a context able to determine that the type argument
```
Comment (gsmet) [inline mapper-pojo/src/main/java/org/hibernate/search/v6poc/entity/pojo/util/impl/GenericTypeContext.java]: This sentence looks unfinished?

File: mapper-pojo/src/main/java/org/hibernate/search/v6poc/entity/pojo/util/impl/GenericTypeContext.java [hunk 2]
```diff
@@ -23,59 +23,139 @@
...
+ * Note that "type resolution" here only covers the resolution of a type parameter to the corresponding argument,
+ * i.e. "shallow" resolution.
+ * This class will not build new {@link ParameterizedType} instances to perform deep resolution.
+ * For instance, given the type {@code ArrayList<? extends Map<String, Integer>>}
+ * When asked
```
Comment (gsmet) [inline mapper-pojo/src/main/java/org/hibernate/search/v6poc/entity/pojo/util/impl/GenericTypeContext.java]: Same here.

File: mapper-pojo/src/main/java/org/hibernate/search/v6poc/entity/pojo/model/spi/PojoCaster.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.v6poc.entity.pojo.model.spi;
+
+public interface PojoCaster<T> {
+
+	T cast(Object object);
```
Comment (gsmet) [inline mapper-pojo/src/main/java/org/hibernate/search/v6poc/entity/pojo/model/spi/PojoCaster.java]: You see, in the end, you want to cast objects ;).

File: mapper-pojo/src/main/java/org/hibernate/search/v6poc/entity/pojo/model/spi/PojoRawTypeModel.java [hunk 4]
```diff
@@ -35,6 +37,14 @@
 	@Override
 	Stream<? extends PojoRawTypeModel<? super T>> getDescendingSuperTypes();
 
+	<A extends Annotation> Optional<A> getAnnotationByType(Class<A> annotationType);
```
Comment (gsmet) [inline mapper-pojo/src/main/java/org/hibernate/search/v6poc/entity/pojo/model/spi/PojoRawTypeModel.java]: In the first sentence of the javadoc of this class, s/An/A/

**Comment by yrodiere** [comment]
I addressed your comments and merged. Thanks!

**Pull request merged**